File: sable-shared/src/config/redis.ts
// ioredis client factory shared by all Sable Node services.
//
// Used for: rate limiting (per-user/org/IP counters), nonce dedup, session
// cache, blocked-entity cache, bot-score counters, route cache. Per the
// gateway DB doc: keys are namespaced (`rate:user:{id}:minute`,
// `nonce:{nonce}`, `session:{id}`, `block:cache:{type}:{value}`, etc.).

import Redis, { type Redis as RedisClient, type RedisOptions } from 'ioredis';

export type { RedisClient };

// Augment ioredis's command interface with our pre-registered scripts. After
// `createRedis()`, every Redis instance has these methods typed and ready —
// no per-call EVAL bytes on the wire (uses EVALSHA with the cached script).
declare module 'ioredis' {
  interface RedisCommander<Context> {
    /**
     * Atomic INCR + EXPIRE-on-first-write. Subsequent INCRs do NOT reset
     * the TTL. Used for fixed-window rate-limit counters.
     * Returns the new counter value.
     */
    sableIncrWithTtl(
      key: string,
      ttlSeconds: number,
    ): import('ioredis').Result<number, Context>;
  }
}

const INCR_WITH_TTL_LUA = `
local count = redis.call('INCR', KEYS[1])
if count == 1 then
  redis.call('EXPIRE', KEYS[1], ARGV[1])
end
return count
`;

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------

export interface RedisConfig {
  /** Logical service name. Surfaces in CLIENT LIST. */
  serviceName: string;
  /**
   * Connection URL — `redis://[:password@]host:port[/db]` or `rediss://...`
   * for TLS. Use this OR explicit host/port, not both.
   */
  url?: string;
  host?: string;
  port?: number;
  password?: string;
  db?: number;
  /** Force TLS even on a plain `redis://` URL — Cloud Memorystore in-transit encryption. */
  tls?: boolean;
  /** Connect timeout in ms. Default 10_000. */
  connectTimeoutMs?: number;
  /**
   * How many times ioredis retries a failed command before throwing.
   * Default 3. Set to null to retry forever (don't — fail fast).
   */
  maxRetriesPerRequest?: number | null;
}

export function createRedis(config: RedisConfig): RedisClient {
  const baseOptions: RedisOptions = {
    connectionName: config.serviceName,
    connectTimeout: config.connectTimeoutMs ?? 10_000,
    maxRetriesPerRequest: config.maxRetriesPerRequest ?? 3,
    enableReadyCheck: true,
    lazyConnect: false,
    // Retry connection with exponential backoff capped at 2s.
    retryStrategy: (times) => Math.min(times * 100, 2_000),
    ...(config.tls ? { tls: {} } : {}),
  };

  const redis = config.url
    ? new Redis(config.url, baseOptions)
    : new Redis({
        host: config.host ?? '127.0.0.1',
        port: config.port ?? 6379,
        password: config.password,
        db: config.db ?? 0,
        ...baseOptions,
      });

  // Pre-register Lua scripts. ioredis caches by SHA and uses EVALSHA —
  // script bytes only ship on first call (or after FLUSHALL/restart).
  redis.defineCommand('sableIncrWithTtl', { numberOfKeys: 1, lua: INCR_WITH_TTL_LUA });

  return redis;
}

// ---------------------------------------------------------------------------
// Lifecycle helpers
// ---------------------------------------------------------------------------

/** Liveness check for `/healthz`. Returns latency in ms or throws. */
export async function pingRedis(redis: RedisClient, timeoutMs = 2000): Promise<number> {
  const start = Date.now();
  let timer: ReturnType<typeof setTimeout> | undefined;
  try {
    await Promise.race([
      redis.ping(),
      new Promise<never>((_, reject) => {
        timer = setTimeout(() => reject(new Error('redis ping timeout')), timeoutMs);
      }),
    ]);
  } finally {
    if (timer !== undefined) clearTimeout(timer);
  }
  return Date.now() - start;
}

/** Graceful shutdown — call on SIGTERM. */
export async function closeRedis(redis: RedisClient): Promise<void> {
  await redis.quit();
}

File: sable-shared-py/sable_shared/config/redis.py
"""Async Redis client factory shared by all Sable Python services.

Used for: rate limiting (per-user/org/IP counters), nonce dedup, session
cache, blocked-entity cache, bot-score counters, route cache. Per the
gateway DB doc: keys are namespaced — every key lives in
``sable_shared.cache.keys``.

``redis`` is an optional dependency of sable-shared — services that
import this module must include ``redis>=5`` in their own requirements.

Mirrors TS ``sable-shared/src/config/redis.ts``.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Awaitable, cast

if TYPE_CHECKING:
    from redis.asyncio import Redis as RedisClient
else:
    try:
        from redis.asyncio import Redis as RedisClient
    except ImportError:  # pragma: no cover — optional dep
        RedisClient = None  # type: ignore[assignment, misc]


# Lua: atomic INCR + EXPIRE-on-first-write. Subsequent INCRs do NOT reset
# the TTL — fixed-window rate limit semantics.
_INCR_WITH_TTL_LUA = """
local count = redis.call('INCR', KEYS[1])
if count == 1 then
  redis.call('EXPIRE', KEYS[1], ARGV[1])
end
return count
"""


@dataclass(frozen=True)
class RedisConfig:
    """Connection config for ``create_redis``.

    Use ``url`` OR explicit host/port, not both.
    """

    service_name: str
    url: str | None = None
    host: str = "127.0.0.1"
    port: int = 6379
    password: str | None = None
    db: int = 0
    tls: bool = False
    connect_timeout: float = 10.0


def create_redis(config: RedisConfig) -> "RedisClient":
    """Build an async Redis client and pre-register the IncrWithTtl script.

    The returned client decodes bytes to ``str`` by default — the Sable
    cache helpers (``set_json`` / ``get_json``) assume string mode.
    """
    if RedisClient is None:
        raise RuntimeError(
            "redis is not installed. Add 'redis>=5' to the service's dependencies "
            "if you need sable_shared.config.redis or sable_shared.cache.",
        )

    # Keep kwargs as Any so mypy strict accepts the **splat against
    # redis's heavily-overloaded constructor signature.
    common: dict[str, Any] = {
        "decode_responses": True,
        "client_name": config.service_name,
        "socket_connect_timeout": config.connect_timeout,
        "health_check_interval": 30,
    }
    if config.tls:
        common["ssl"] = True

    if config.url is not None:
        client = RedisClient.from_url(config.url, **common)
    else:
        client = RedisClient(
            host=config.host,
            port=config.port,
            password=config.password,
            db=config.db,
            **common,
        )

    # Attach the pre-registered script as ``client.sable_incr_with_ttl`` so
    # the cache helpers can call it without re-registering on every call.
    # ``register_script`` returns a callable that uses EVALSHA + falls back
    # to EVAL on NOSCRIPT.
    client.sable_incr_with_ttl = client.register_script(_INCR_WITH_TTL_LUA)  # type: ignore[attr-defined]

    return client


async def ping_redis(client: "RedisClient", timeout: float = 2.0) -> float:
    """Liveness check for ``/healthz``. Returns latency in ms or raises."""
    loop = asyncio.get_running_loop()
    start = loop.time()
    # `ping()` is typed as `Awaitable[bool] | bool` (the sync-cluster path);
    # in the asyncio client it's always awaitable — cast to keep mypy happy.
    async with asyncio.timeout(timeout):
        await cast(Awaitable[Any], client.ping())
    return (loop.time() - start) * 1000


async def close_redis(client: "RedisClient") -> None:
    """Graceful shutdown — call on SIGTERM."""
    await client.aclose()
